added mysat1 telemetry and telemetry parser

Hamze-Issa opened the pull request:
I added mysat1 telemetry and telemetry parser to the python folder. Not sure if that's all there is to do to build the blocks though, so let me know if anything else is needed or if there is anything wrong in the added files themselves.
Thank you

daniestevez commented on python/mysat1_telemetry_parser.py:
```diff
@@ -0,0 +1,53 @@
+        packet = bytearray(pmt.u8vector_elements(msg))
+
+        try:
+            data = mysat1_telemetry.Beacon.parse(packet[4:])
```
Probably you need to skip 16 bytes rather than 4 to get to the packet's payload.

daniestevez commented:
Please include the parser as a complete GNU Radio block, with an XML file in the grc/ folder, appropriate entries in CMakeLists.txt and __init__.py, add the parser to mysat1.grc, and test it. I have spotted one error from reviewing the code, and this error should have been caught in testing.

Hamze-Issa commented:
Check it now, I tested it by receiving the actual telemetry from the satellite with it. I'm new to GNURadio so I didn't know what to do to build the block, but now it should work nicely.

daniestevez commented:
A couple of notes:
* "Unit" conversion (dividing voltages by 10, etc) should be done with Construct. Take a look at `adapters.py` and how this is used elsewhere.
* There are some unnecessary changes in `mysat.grc` and spurious files (see code review).

daniestevez commented on apps/mysat1.grc:
```diff
@@ -624,7 +612,7 @@
     <param>
       <key>ipaddr</key>
-      <value>ip</value>
+      <value></value>
```
Unwanted change

daniestevez commented on apps/mysat1.grc:
```diff
@@ -648,7 +636,7 @@
     <param>
       <key>port</key>
-      <value>port</value>
+      <value>3000</value>
```
Unwanted change

Hamze-Issa commented:
I restored the original mysat1.grc file with only the telemetry_parser block added. Removed the .pyc file and put all the conversions in the Construct. I also tested the new conversion.

daniestevez commented:
Thanks for the work. The code looks correct now. However, I was testing it with the sample packet in `satellite-recordings/mysat1.wav` and I get very weird values for most of the channels:

```Container: 
    callsign = Container: 
        callsign = ListContainer: 
            65
            54
            56
            77
            89
    obc = Container: 
        obc_mode = 2
        obc_reset_counter = 369098752
        obc_uptime = 998508032
    gyro = Container: 
        gyro_norm = 3
    timestamp = Container: 
        timestamp = 2018-05-20 04:56:32
    obc_temp = Container: 
        obc_temp = -128.0
        obc_daughter_board_temp = -125.0
    eps_temp = Container: 
        eps_battery_temp = -56.0
        eps_board_temp = -101.0
    ants = Container: 
        ants_temp = 59.0
    trxvu_temp = Container: 
        trxvu_temp = -36.0
    adcs = Container: 
        adcs_temp = 9.0
    solar_panels = Container: 
        solar_panels_temp = ListContainer: 
            12.0
            7.0
            7.0
            10.0
            9.0
            12.0
    obc_voltages = Container: 
        obc_3v3_voltage = 3.3
        obc_5v0_voltage = 0.0
    trxvu_voltage = Container: 
        trxvu_voltage = 8.2
    eps_batt_voltage = Container: 
        eps_batt_voltage = 8.3
    obc_current = Container: 
        obc_5.0_current = 0
    eps_currents = Container: 
        eps_total_pv_current = 52480
        eps_total_system_current = 36608
```

The timestamp is in the future, `eps_currents` are too large (I think they're supposed to be in mA, right?), OBC and EPS temperatures look too low. Is the satellite transmitting these incorrect values or is this a problem with parameter conversion?

daniestevez commented on python/mysat1_telemetry.py:
```diff
@@ -0,0 +1,95 @@
+Timestamp = UNIXTimestampAdapter(Int32ub)
+
+Callsign = Struct(
+    'callsign' /  Int8ub[5])
```
It would be better to use `Bytes(5)` instead of `Int8ub[5]`.

Hamze-Issa replied:
ok, will do

Hamze-Issa commented:
I checked with the MySat1 team and It turns out the document I was using (the one I sent you) had missing values. Also, the values that had more than one byte need to be reversed (C language little endian). So I got the missing values from the MySat1 team and will integrate it in the code. I'll also fix the reversing issue. If you did the reversing part in your previous satellite decoders I'll appreciate it if you guide me to an example so I can make it as compatible with your code as possible. Thank you

daniestevez commented:
What do you mean by "reversing"?

Hamze-Issa commented:
The byte swapping. The construct reads the data most significant byte first (big endian), but it's stored to be least significant byte first (little endian). That's why we need to swap the bytes of values that are constructed of more than one byte.

daniestevez commented:
In Construct you have different datatypes to handle endianness. For example, you have `Int16sb` for an `int16_t` in big-endian format and `Int16sl` for an `int16_t` in little-endian format. Thus, it is not necessary to swap bytes. You just need to use the appropriate datatype.

Hamze-Issa commented:
Aha. OK great, thanks...

Hamze-Issa commented:
I added 3 missing values (eps_counter_boot,  eps_cause_last_boot, eps_battery_mode) and removed a deprecated value (solar panels temp). I also tested with the Mysat1 engineering model and got these values:

> 2019-04-30 07:58:51
Packet number 1
Container: 
    callsign = Container: 
        callsign = b'A68MY' (total 5)
    obc = Container: 
        obc_mode = 1
        obc_reset_counter = 156
        obc_uptime = 305
    gyro = Container: 
        gyro_norm = 1
    eps = Container: 
        eps_counter_boot = 1051
        eps_cause_last_boot = 7
        eps_battery_mode = 3
    timestamp = Container: 
        timestamp = 2019-04-01 04:55:33
    obc_temp = Container: 
        obc_temp = 28.0
        obc_daughter_board_temp = 28.0
    eps_temp = Container: 
        eps_battery_temp = 26.0
        eps_board_temp = 27.0
    ants = Container: 
        ants_temp = 28.0
    trxvu_temp = Container: 
        trxvu_temp = 28.0
    adcs = Container: 
        adcs_temp = 28.0
    obc_voltages = Container: 
        obc_3v3_voltage = 3.3
        obc_5v0_voltage = 0.2
    trxvu_voltage = Container: 
        trxvu_voltage = 7.4
    eps_batt_voltage = Container: 
        eps_batt_voltage = 7.4
    obc_current = Container: 
        obc_5.0_current = 0.0
    eps_currents = Container: 
        eps_total_pv_current = 0.007
        eps_total_system_current = 0.141

daniestevez commented:
Thank you so much for your work. Now I think everything is correct. I have merge the pull request.
